srtp/srtp.c

```diff
@@ -1727,10 +1731,29 @@ static srtp_err_status_t srtp_protect_aead(srtp_ctx_t *ctx,
      * extension, if present; otherwise, it starts after the last csrc,
      * if any are present
      */
-    enc_start = (uint32_t *)hdr + uint32s_in_rtp_header + hdr->cc;
+
+    /* Cryptex can only encrypt CSRCS if header extension is present*/
+    if (stream->use_cryptex && hdr->cc && !hdr->x) {
+        return srtp_err_status_parse_err;
+    }
     if (hdr->x == 1) {
-        xtn_hdr = (srtp_hdr_xtnd_t *)enc_start;
-        enc_start += (ntohs(xtn_hdr->length) + 1);
+        xtn_hdr = (srtp_hdr_xtnd_t *)(uint32_t *)hdr + uint32s_in_rtp_header +
+                  hdr->cc;
+        xtn_hdr_length = ntohs(xtn_hdr->length);
+        xtn_profile_specific = ntohs(xtn_hdr->profile_specific);
+    }
+    /* If no header extension is present cryptex has no effect */
+    if (stream->use_cryptex && hdr->x) {
+        uint32_t *csrcs = (uint32_t *)hdr + uint32s_in_rtp_header;
+        /* Move CSRCS so it is contiguos with extension header block */
+        for (unsigned char i = 0; i < hdr->cc; ++i)
+            csrcs[i + 1] = csrcs[i];
+        enc_start = csrcs + 1;
+    } else {
+        enc_start = (uint32_t *)hdr + uint32s_in_rtp_header + hdr->cc;
+        if (hdr->x == 1) {
+            enc_start += (xtn_hdr_length + 1);
+        }
```

Comment:
Am I missing something, or does this not do AAD over the header extension header (the "defined by profile" and length fields) if CSRCs are present?  I think the first CSRC ends up being put in the AAD instead.

Reply:
Yes, that is incorrect. First CSRC should be overwritten by "defined by profile" and length fields so they are included on the AAD, right?

Reply:
Yes, and then that field should be put back in its proper place when the CSRCs are moved back.

Reply:
Moving back the CSRCs and restoring the defined by profile + len is done after encryption below. I will have to replace the 0xBEDE and 0x100X by the crytpex  one here so they correct values are used in the authentication. Currently it is done after authentication.

Reply:
should be fixed now